hathor/sysctl/core/manager.py

```diff
@@ -61,3 +93,141 @@ def set_profiler_stop(self, save_to: str | None) -> None:
         if not save_to:
             save_to = None
         self.manager.stop_profiler(save_to=save_to)
+
+    @signal_handler_safe
+    def set_pudb_set_trace_attach_tty(self, tty: str) -> None:
+        """Stop execution and open pudb in a given tty.
+
+        ATTENTION: This command can be destructive and the full node might not work after running it.
+
+        Open a new terminal. First, you need to get the path of the tty of the
+        terminal you want to debug from. To do that, use the standard unix
+        command `tty`. It will print something like `/dev/pts/3`.
+
+        Then you need to make sure that your terminal doesn't have a shell actively
+        reading and possibly capturing some of the input that should go to pudb.
+        To do that run a placeholder command that does nothing, such as `perl -MPOSIX -e pause`.
+        """
+        fp = open(tty, 'r+b', buffering=0)
+        term_size = os.get_terminal_size(fp.fileno())
+        self._run_pudb_set_trace(tty, fp, term_size=term_size)
+
+    @signal_handler_safe
+    def set_pudb_set_trace_create_tty(self, cols: int, rows: int) -> None:
+        """Stop execution and open pudb for debugging in a newly created tty.
+
+        ATTENTION: This command can be destructive and the full node might not work after running it.
+
+        You must provide the terminal size (cols, rows).
+
+        The newly created tty path will be printed in the logs. After you get it, you can
+        connect to it using `screen <ttyname>`.
+        """
+        import fcntl
+        import struct
+        import termios
+
+        # Some of these methods are not available in some operating system (e.g. Windows).
+        # This is a way to avoid mypy errors.
+        openpty = getattr(os, 'openpty', None)
+        ioctl = getattr(fcntl, 'ioctl', None)
+        TIOCSWINSZ = getattr(termios, 'TIOCSWINSZ', None)
+
+        if openpty is None or ioctl is None or TIOCSWINSZ is None:
+            self.log.warn('Error opening pudb', openpty=openpty, ioctl=ioctl, TIOCSWINSZ=TIOCSWINSZ)
+            return
+
+        (term_master, term_slave) = openpty()
+
+        term_size = (cols, rows)
+        term_size_bytes = struct.pack("HHHH", term_size[1], term_size[0], 0, 0)
+        ioctl(term_master, TIOCSWINSZ, term_size_bytes)
+
+        tty_name = os.ttyname(term_slave)
+        fp = os.fdopen(term_master, 'wb+', buffering=0)
+
+        self._run_pudb_set_trace(tty_name, fp, term_size=term_size)
+
+    def _run_pudb_set_trace(self, tty: str, fp: IO[bytes], *, term_size: Optional[tuple[int, int]] = None) -> None:
+        from pudb.debugger import Debugger
+
+        self.log.warn('main loop paused; pudb.set_trace running', tty=tty)
+
+        tty_file = io.TextIOWrapper(fp)
+        kwargs: dict[str, Any] = {
+            'stdin': tty_file,
+            'stdout': tty_file,
+        }
+        if term_size is not None:
```

Comment:
If `term_size` is `None` what happens? Does it get a default value?

Reply:
It tries to get the term size from the envvars and using ioctl. This is the default behavior from pudb.